File: scripts/enrich-lead-phones.mjs
/**
 * Official-website phone enrichment for all five supported industries.
 * Default is dry-run. Use --apply to update active_leads.
 */
import fs from "node:fs";

const INDUSTRIES = [
  "Automobile & Auto Components",
  "Pharmaceuticals & Healthcare Manufacturing",
  "Chemical Manufacturing & Allied Industries",
  "Packaging, Plastics & Paper Manufacturing",
  "Food Processing & Agro Manufacturing",
];
const BLOCKED_HOSTS = ["indiamart.com", "tradeindia.com", "justdial.com", "exportersindia.com", "facebook.com", "instagram.com", "linkedin.com", "page.link", "bit.ly", "tinyurl.com", "t.co", "goo.gl"];
const PHONE_PATTERN = /(?<!\d)(?:\+?91[\s.()/-]*)?(?:0?[6-9]\d{9}|0?[1-9]\d{7,10})(?!\d)/g;
const apply = process.argv.includes("--apply");
const limitIndex = process.argv.indexOf("--limit-per-industry");
const perIndustryLimit = limitIndex >= 0 ? Math.max(1, Number(process.argv[limitIndex + 1]) || 100) : 100;

function loadEnv() {
  for (const file of [".env.local", ".env"]) {
    if (!fs.existsSync(file)) continue;
    for (const line of fs.readFileSync(file, "utf8").split(/\r?\n/)) {
      if (!line || line.trimStart().startsWith("#") || !line.includes("=")) continue;
      const [key, ...rest] = line.split("=");
      if (!process.env[key.trim()]) process.env[key.trim()] = rest.join("=").trim().replace(/^['"]|['"]$/g, "");
    }
  }
}

function host(url) {
  try { return new URL(url).hostname.toLowerCase().replace(/^www\./, ""); } catch { return ""; }
}
function officialWebsite(url) {
  try {
    const parsed = new URL(url);
    const name = host(url);
    return ["http:", "https:"].includes(parsed.protocol) && name && !BLOCKED_HOSTS.some((domain) => name === domain || name.endsWith(`.${domain}`));
  } catch { return false; }
}
function normalizePhone(raw) {
  let digits = String(raw).replace(/\D/g, "");
  if (digits.startsWith("91") && digits.length > 10) digits = digits.slice(2);
  digits = digits.replace(/^0+/, "");
  if (digits.length === 10 && /^[6-9]/.test(digits)) return `+91 ${digits}`;
  if (digits.length >= 8 && digits.length <= 11) return `0${digits}`;
  return null;
}
function extractPhones(html) {
  const candidates = [...html.matchAll(/href\s*=\s*["']tel:([^"']+)/gi)].map((m) => m[1]);
  candidates.push(...(html.replace(/<[^>]*>/g, " ").match(PHONE_PATTERN) || []));
  return [...new Set(candidates.map(normalizePhone).filter(Boolean))];
}
async function loadPage(url) {
  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), 3000);
  try {
    const response = await fetch(url, { headers: { "user-agent": "LeadFlow official contact verifier" }, redirect: "follow", signal: controller.signal });
    if (!response.ok || host(response.url) !== host(url)) return null;
    return { url: response.url, html: await response.text() };
  } catch { return null; } finally { clearTimeout(timer); }
}
async function fetchWithRetry(url, options, attempts = 3) {
  let lastError;
  for (let attempt = 1; attempt <= attempts; attempt += 1) {
    try { return await fetch(url, options); } catch (error) { lastError = error; }
  }
  throw lastError;
}
function contactUrls(homeUrl, html) {
  const values = [];
  const regex = /<a\b[^>]*href\s*=\s*["']([^"']+)["'][^>]*>([\s\S]*?)<\/a>/gi;
  for (const match of html.matchAll(regex)) {
    const label = `${match[1]} ${match[2].replace(/<[^>]*>/g, " ")}`.toLowerCase();
    if (!/(contact|contact-us|contactus|reach-us|get-in-touch|connect)/.test(label)) continue;
    try {
      const candidate = new URL(match[1], homeUrl).toString();
      if (host(candidate) === host(homeUrl)) values.push(candidate);
    } catch { /* ignore malformed URL */ }
  }
  return [...new Set(values)].slice(0, 2);
}
async function websitePhones(url) {
  const homepage = await loadPage(url);
  if (!homepage) return [];
  const numbers = extractPhones(homepage.html);
  for (const contactUrl of contactUrls(homepage.url, homepage.html)) {
    const contactPage = await loadPage(contactUrl);
    if (contactPage) numbers.push(...extractPhones(contactPage.html));
  }
  return [...new Set(numbers)];
}

loadEnv();
const supabaseUrl = (process.env.NEXT_PUBLIC_SUPABASE_URL || "").replace(/\/$/, "");
const serviceKey = process.env.SUPABASE_SERVICE_ROLE_KEY || "";
if (!supabaseUrl || !serviceKey) throw new Error("Missing Supabase environment variables.");
const apiUrl = `${supabaseUrl}/rest/v1/active_leads`;
const headers = { apikey: serviceKey, Authorization: `Bearer ${serviceKey}`, "Content-Type": "application/json" };

let scanned = 0;
let candidates = 0;
let updated = 0;
for (const industry of INDUSTRIES) {
  const query = new URLSearchParams({ select: "id,company_name,industry,phone,website", industry: `eq.${industry}`, limit: String(perIndustryLimit) });
  const response = await fetchWithRetry(`${apiUrl}?${query}`, { headers });
  if (!response.ok) throw new Error(`Supabase read failed for ${industry}: ${response.status}`);
  const leads = await response.json();
  for (const lead of leads) {
    scanned += 1;
    if (!officialWebsite(lead.website)) continue;
    const numbers = await websitePhones(lead.website);
    const existing = normalizePhone(lead.phone);
    const websitePhone = numbers.includes(existing) ? existing : numbers.find((number) => number.startsWith("+91 ")) || numbers[0];
    if (!websitePhone || websitePhone === existing) continue;
    candidates += 1;
    console.log(`${apply ? "UPDATE" : "WOULD UPDATE"} | ${lead.company_name} | ${lead.phone || "N/A"} -> ${websitePhone}`);
    if (!apply) continue;
    const digits = websitePhone.replace(/\D/g, "").replace(/^91/, "");
    const mobile = /^[6-9]\d{9}$/.test(digits);
    const patch = await fetchWithRetry(`${apiUrl}?id=eq.${encodeURIComponent(lead.id)}`, {
      method: "PATCH",
      headers: { ...headers, Prefer: "return=minimal" },
      body: JSON.stringify({ phone: websitePhone, phone_type: mobile ? "Mobile" : "Landline", is_whatsapp: mobile, whatsapp_link: mobile ? `https://wa.me/91${digits}` : null }),
    });
    if (!patch.ok) throw new Error(`Update failed for ${lead.company_name}: ${patch.status}`);
    updated += 1;
  }
}
console.log(apply
  ? `Updated ${updated} of ${candidates} verified candidates after scanning ${scanned} leads.`
  : `Dry-run found ${candidates} verified candidates after scanning ${scanned} leads.`);

File: scripts/enrich_lead_phones.py
"""Enrich active_leads with phone numbers found on official company websites.

Rules:
1. Use a phone found on the company domain (homepage or Contact Us page) first.
2. Never treat marketplace or URL-shortener links as an official website.
3. If no website phone is found, retain the existing Maps/legacy number as fallback.

The command is read-only by default. Pass --apply only after reviewing its output.
"""

from __future__ import annotations

import argparse
import os
import re
import sys
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Iterable
from urllib.parse import urljoin, urlparse

import requests
from bs4 import BeautifulSoup

try:
    from supabase import create_client
except ImportError as error:  # pragma: no cover - operational guard
    raise SystemExit("Install python dependencies first: pip install -r requirements.txt") from error


ALL_INDUSTRIES = [
    "Automobile & Auto Components",
    "Pharmaceuticals & Healthcare Manufacturing",
    "Chemical Manufacturing & Allied Industries",
    "Packaging, Plastics & Paper Manufacturing",
    "Food Processing & Agro Manufacturing",
]
MARKETPLACE_HOSTS = {
    "indiamart.com", "tradeindia.com", "justdial.com", "exportersindia.com",
    "facebook.com", "instagram.com", "linkedin.com",
}
SHORT_LINK_HOSTS = {"page.link", "bit.ly", "tinyurl.com", "t.co", "goo.gl"}
CONTACT_WORDS = ("contact", "contact-us", "contactus", "get-in-touch", "reach-us", "connect")
PHONE_PATTERN = re.compile(r"(?<!\d)(?:\+?91[\s.()-]*)?(?:0?[6-9]\d{9}|0?[1-9]\d{7,10})(?!\d)")
HEADERS = {"User-Agent": "LeadFlow-Data-Validator/1.0 (+website-contact-verification)"}


def load_local_env() -> None:
    """Load local development variables without adding a dotenv dependency."""
    for filename in (".env.local", ".env"):
        path = Path(filename)
        if not path.exists():
            continue
        for line in path.read_text(encoding="utf-8").splitlines():
            if not line or line.lstrip().startswith("#") or "=" not in line:
                continue
            key, value = line.split("=", 1)
            os.environ.setdefault(key.strip(), value.strip().strip('"').strip("'"))


def normalized_host(url: str) -> str:
    return urlparse(url).hostname.lower().removeprefix("www.") if urlparse(url).hostname else ""


def is_official_website(url: str | None) -> bool:
    if not url:
        return False
    parsed = urlparse(url)
    host = normalized_host(url)
    if parsed.scheme not in {"http", "https"} or not host:
        return False
    blocked = MARKETPLACE_HOSTS | SHORT_LINK_HOSTS
    return not any(host == domain or host.endswith(f".{domain}") for domain in blocked)


def normalize_indian_phone(raw: str) -> str | None:
    digits = re.sub(r"\D", "", raw)
    if digits.startswith("91") and len(digits) > 10:
        digits = digits[2:]
    digits = digits.lstrip("0")
    if len(digits) == 10 and digits[0] in "6789":
        return f"+91 {digits}"
    if 8 <= len(digits) <= 11:
        return f"0{digits}"
    return None


def extract_phones(soup: BeautifulSoup) -> list[str]:
    values: list[str] = []
    for tag in soup.select('a[href^="tel:"]'):
        value = normalize_indian_phone(tag.get("href", "").removeprefix("tel:"))
        if value:
            values.append(value)
    visible_text = soup.get_text(" ", strip=True)
    for match in PHONE_PATTERN.findall(visible_text):
        value = normalize_indian_phone(match)
        if value:
            values.append(value)
    return list(dict.fromkeys(values))


def contact_page_urls(home_url: str, soup: BeautifulSoup) -> Iterable[str]:
    home_host = normalized_host(home_url)
    for anchor in soup.find_all("a", href=True):
        href = anchor["href"].strip()
        label = f"{anchor.get_text(' ', strip=True)} {href}".lower()
        candidate = urljoin(home_url, href)
        if normalized_host(candidate) == home_host and any(word in label for word in CONTACT_WORDS):
            yield candidate


def fetch_website_phones(url: str, timeout: int) -> tuple[list[str], str]:
    try:
        response = requests.get(url, headers=HEADERS, timeout=timeout, allow_redirects=True)
        response.raise_for_status()
        if normalized_host(response.url) != normalized_host(url):
            return [], "website redirected to another domain"
        homepage = BeautifulSoup(response.text, "html.parser")
        phones = extract_phones(homepage)
        checked = {response.url}
        for contact_url in list(dict.fromkeys(contact_page_urls(response.url, homepage)))[:2]:
            if contact_url in checked:
                continue
            checked.add(contact_url)
            try:
                contact_response = requests.get(contact_url, headers=HEADERS, timeout=timeout, allow_redirects=True)
                contact_response.raise_for_status()
                if normalized_host(contact_response.url) == normalized_host(url):
                    phones.extend(extract_phones(BeautifulSoup(contact_response.text, "html.parser")))
            except requests.RequestException:
                continue
        unique = list(dict.fromkeys(phones))
        return unique, "website contact found" if unique else "no phone on homepage/contact page"
    except requests.RequestException as error:
        return [], f"website unavailable: {error.__class__.__name__}"


def choose_website_phone(candidates: list[str], existing_phone: str | None) -> str | None:
    if not candidates:
        return None
    existing = normalize_indian_phone(existing_phone or "")
    if existing and existing in candidates:
        return existing
    # Prefer a mobile number when a company exposes both a mobile and a landline.
    return next((phone for phone in candidates if phone.startswith("+91 ")), candidates[0])


def fetch_rows(client: Any, industries: list[str], limit: int) -> list[dict[str, Any]]:
    rows: list[dict[str, Any]] = []
    for industry in industries:
        result = client.table("active_leads").select(
            "id,company_name,industry,phone,phone_type,is_whatsapp,whatsapp_link,website,location,city"
        ).eq("industry", industry).limit(limit).execute()
        rows.extend(result.data or [])
    return rows


def main() -> int:
    parser = argparse.ArgumentParser(description="Verify stored lead phones against official websites.")
    parser.add_argument("--apply", action="store_true", help="Write reviewed enrichment results to Supabase.")
    parser.add_argument("--limit-per-industry", type=int, default=100, help="Maximum rows to inspect for each of the five industries.")
    parser.add_argument("--timeout", type=int, default=10, help="Website request timeout in seconds.")
    parser.add_argument("--industry", action="append", choices=ALL_INDUSTRIES, help="Run for one industry; repeat to select more.")
    args = parser.parse_args()

    load_local_env()
    url = os.environ.get("NEXT_PUBLIC_SUPABASE_URL", "").rstrip("/")
    key = os.environ.get("SUPABASE_SERVICE_ROLE_KEY", "")
    if not url or not key:
        print("Missing NEXT_PUBLIC_SUPABASE_URL or SUPABASE_SERVICE_ROLE_KEY.", file=sys.stderr)
        return 2

    client = create_client(url, key)
    industries = args.industry or ALL_INDUSTRIES
    rows = fetch_rows(client, industries, args.limit_per_industry)
    print(f"Inspecting {len(rows)} leads across {len(industries)} industries ({'APPLY' if args.apply else 'DRY RUN'}).")

    updates = 0
    for row in rows:
        website = row.get("website")
        existing_phone = row.get("phone")
        if not is_official_website(website):
            print(f"KEEP  | {row['company_name']} | maps fallback | no official website")
            continue

        candidates, reason = fetch_website_phones(website, args.timeout)
        website_phone = choose_website_phone(candidates, existing_phone)
        if not website_phone:
            print(f"KEEP  | {row['company_name']} | maps fallback | {reason}")
            continue

        current_normalized = normalize_indian_phone(existing_phone or "")
        changed = current_normalized != website_phone
        print(f"{'UPDATE' if changed else 'MATCH '} | {row['company_name']} | {website_phone} | {reason}")
        if not args.apply or not changed:
            continue

        digits = re.sub(r"\D", "", website_phone).removeprefix("91")
        is_mobile = len(digits) == 10 and digits[0] in "6789"
        payload = {
            "phone": website_phone,
            "phone_type": "Mobile" if is_mobile else "Landline",
            "is_whatsapp": is_mobile,
            "whatsapp_link": f"https://wa.me/91{digits}" if is_mobile else None,
            "phone_source": "official_website",
            "phone_verified_at": datetime.now(timezone.utc).isoformat(),
        }
        client.table("active_leads").update(payload).eq("id", row["id"]).execute()
        updates += 1

    print(f"Completed. {updates} rows updated." if args.apply else "Completed dry run. No database rows were changed.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
